crates/sui-data-ingestion/src/worker_pool.rs

```diff
@@ -0,0 +1,108 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::executor::MAX_CHECKPOINTS_IN_PROGRESS;
+use crate::workers::Worker;
+use mysten_metrics::spawn_monitored_task;
+use std::collections::HashSet;
+use std::sync::Arc;
+use sui_types::full_checkpoint_content::CheckpointData;
+use sui_types::messages_checkpoint::CheckpointSequenceNumber;
+use tokio::sync::oneshot;
+use tokio::sync::{broadcast, mpsc};
+use tracing::info;
+
+pub struct WorkerPool<W: Worker> {
+    pub task_name: String,
+    concurrency: usize,
+    worker: Arc<W>,
+}
+
+impl<W: Worker + 'static> WorkerPool<W> {
+    pub fn new(worker: W, concurrency: usize) -> Self {
+        Self {
+            task_name: worker.name().to_string(),
+            concurrency,
+            worker: Arc::new(worker),
+        }
+    }
+    pub async fn run(
+        self,
+        mut current_checkpoint_number: CheckpointSequenceNumber,
+        mut checkpoint_receiver: broadcast::Receiver<CheckpointData>,
+        executor_progress_sender: mpsc::Sender<(String, CheckpointSequenceNumber)>,
+    ) {
+        info!(
+            "Starting indexing pipeline {} with concurrency {}",
+            self.task_name, self.concurrency
+        );
+        let mut updates: HashSet<u64> = HashSet::new();
+
+        let (progress_sender, mut progress_receiver) = mpsc::channel(MAX_CHECKPOINTS_IN_PROGRESS);
+        let mut workers = vec![];
+
+        // spawn child workers
+        for _ in 0..self.concurrency {
```

Comment:
@gegaowp: let me know if such concurrency works or do you expect some perf degradation comparing to the current indexer implementation

Reply:
wdyt if one worker crashed the binary, what's the trade offs betwen one binary with spawned workers, and configurable separate binaries, assuming that we can maintain the states / worker metadata in S3?

Reply:
the progress state is stored in an external store (for instance, DynamoDB). The daemon is configurable, allowing us to experiment with either spawning all workflows in one daemon or utilizing a couple of instances